scripts/odbpy/power_utils.py

```diff
@@ -138,149 +139,126 @@ def find_power_ground_port(port_name, ports):
                 cell_name = cell_name
                 print(
-                    "Warning: power pin",
-                    pin_name,
-                    "of",
-                    cell_name,
-                    "is already connected",
+                    f"[WARN] {cell_name}/{pin_name} appears to already be connected. Ignoring..."
                 )
-                print("Warning: ignoring", cell_name + "/" + pin_name, "!!!!!!!")
             else:
-                VDD_ITERM.connect(DEFAULT_VDD)
+                VDD_ITERM.connect(vdd_net)
 
         for GND_ITERM in GND_ITERMS:
             if GND_ITERM.isConnected():
                 pin_name = GND_ITERM.getMTerm().getName()
                 cell_name = cell_name
                 print(
-                    "Warning: ground pin",
-                    pin_name,
-                    "of",
-                    cell_name,
-                    "is already connected",
+                    f"[WARN] {cell_name}/{pin_name} appears to already be connected. Ignoring..."
                 )
-                print("Warning: ignoring", cell_name + "/" + pin_name, "!!!!!!!")
             else:
-                GND_ITERM.connect(DEFAULT_GND)
+                GND_ITERM.connect(gnd_net)
 
         modified_cells += 1
 
-    print(
-        "Modified power connections of",
-        modified_cells,
-        "cells (Remaining:",
-        len(cells) - modified_cells,
-        ").",
-    )
+    print(f"Modified power connections of {modified_cells}/{len(cells)} cells.")
 
     # apply extra special connections taken from another netlist:
+    ### ^ Literally what in God's name does this mean?
     if powered_netlist is not None and os.path.exists(powered_netlist):
-        tmp_def_file = f"{os.path.splitext(powered_netlist)[0]}.def"
-        openroad_script = []
-        openroad_script.append(f"read_lef {input_lef}")
-        openroad_script.append(f"read_verilog {powered_netlist}")
-        openroad_script.append(f"link_design {reader.name}")
-        openroad_script.append(f"write_def {tmp_def_file}")
-        openroad_script.append("exit")
+        tmp_def_file = f"{os.path.splitext(powered_netlist)[0]}.intermediate.def"
 
-        p = Popen(["openroad"], stdout=PIPE, stdin=PIPE, stderr=PIPE, encoding="utf8")
+        openroad_script = f"""
+        read_lef {input_lef}
+        read_verilog {powered_netlist}
+        link_design {reader.name}
+        write_def {tmp_def_file}
+        exit
+        """
 
-        openroad_script = "\n".join(openroad_script)
-
-        openroad_stdout, openroad_stderr = p.communicate(openroad_script)
-        print(f"STDOUT: {openroad_stdout.strip()}")
-        print(f"STDERR: {openroad_stderr.strip()}")
-        print("openroad exit code:", p.returncode)
-        assert p.returncode == 0, p.returncode
-        assert os.path.exists(tmp_def_file), "DEF file doesn't exist"
+        subprocess.run(["openroad"], check=True, input=openroad_script.encode("utf8"))
```

Comment:
subprocess.run won't exit the main script if the process fails right?

Reply:
With `check=True`, it will!